### ISSUE TITLE
Plan/Apply error: Cannot register provider Microsoft.TimeSeriesInsights

### ISSUE THREAD

**Issue by hashem-alrifai** [hashicorp/terraform-provider-azurerm#27466]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

Latest, i.e. 1.9

### AzureRM Provider Version

3.72.0

### Affected Resource(s)/Data Source(s)

azurerm_provider

### Terraform Configuration Files

```hcl
-
```

### Debug Output/Panic Output

```shell
Original Error: Cannot register providers: Microsoft.TimeSeriesInsights. Errors were: Cannot register provider Microsoft.TimeSeriesInsights with Azure Resource Manager: unexpected status 400 with error: DisallowedProvider: The operation is not permitted for namespace 'Microsoft.TimeSeriesInsights'
```

### Expected Behaviour

No errors. 

### Actual Behaviour

Suddenly today, we started to get the following error in all our Terraform jobs. 
`Original Error: Cannot register providers: Microsoft.TimeSeriesInsights. Errors were: Cannot register provider Microsoft.TimeSeriesInsights with Azure Resource Manager: unexpected status 400 with error: DisallowedProvider: The operation is not permitted for namespace 'Microsoft.TimeSeriesInsights'`

After googling, I knew that Microsoft deprecated the provider! And my questions are: 

- I don't believe we use any resource of that provider! why do I see this error? How can I query if I have a resource under that provider?
- Can I fix all the jobs without needing to upgrade TF version in all repos (a lot). 
- Why can't I see any deprecation warning, so I could upgrade in a good time? 

Thanks.

**Comment by gfosdike1** [comment]
I also have this issue:

azurerm - 3.89
Terraform Version: 1.8.0

**Comment by Roberto-Itauros** [comment]
I have this issue too.

azurerm - 3.86.0
Terraform Version: 1.8.5

(Already reacted with 👍 )

**Comment by markphillipsbjss** [comment]
We've been unable to reproduce this issue after bumping the provider version to azurerm 3.116.0

**Comment by nichollsc81** [comment]
Also seeing this behaviour with provider version `3.75`.

Interestingly reading over debug output, it appears that the provider is trying to parse what looks like a tag etc. being returned from the API now:
```
{...,"@timestamp":"2024-09-23T12:13:23.386123+01:00"}

2024-09-23T12:13:24.977+0100 [DEBUG] provider.terraform-provider-azurerm_v3.75.0_x5: Determining which Resource Providers require Registration: timestamp=2024-09-23T12:13:24.977+0100
```

**Comment by vanchogeorgievski** [comment]
I'm experiencing the same issue

azurerm v3.70.0
terraform both versions 1.5.5 and 1.9.6

**Comment by reespozzi** [comment]
According to MS the retirement of this `Microsoft.TimeSeriesInsights` was brought forward https://azure.microsoft.com/en-us/updates/we-re-retiring-azure-time-series-insights-on-7-july-2024-transition-to-azure-data-explorer/ - and potentially decommissioned first and then quietly removed today... hence sudden faillures
We can get around this by setting `skip_provider_registration = "true"` but this is a deprecated argument for azurerm and to be removed >v4

**Comment by hashem-alrifai** [comment]
So, looks like there is no escape from upgrading the azurerm provider in all my repos 😫

**Comment by ahendershex** [comment]
Same issue. Annoyingly we are stuck on an older version of azurerm until a bug is fixed in latest! At least the skip_provider_registration = "true" workaround is available. for now....

**Comment by Mohapatraabhishek-lab** [comment]
for me upgrading to v 3.112.0 works

**Comment by harshavmb** [comment]
This issue is observed in versions prior to `3.89.0`. Thanks to [#24571](https://github.com/hashicorp/terraform-provider-azurerm/pull/24571), atleast it's going to work from `3.90.0` if your code isn't read for `4.x` versions.

**Comment by kngjaime** [comment]
Having the same issue with 3.85.0. 

I've opened a case number with Microsoft since they may have removed the resource provider: Microsoft.TimeSeriesInsight , and terraform is trying to validate if the provider is registered

**Comment by vromanenko-abs** [comment]
For me upgrading to 3.90.0 works.

**Comment by anemiroffsumer** [comment]
Getting this issue to as of 5 minutes ago when running a plan via ENV0

**Comment by AndreasMWalter** [comment]
> This issue is observed in versions prior to `3.89.0`. Thanks to [#24571](https://github.com/hashicorp/terraform-provider-azurerm/pull/24571), at least it's going to work from `3.90.0` if your code isn't ready for `4.x` versions.

Yeah, no dice for me using 3.116 :/ looking into a quick fix now though, will report back
![image](https://github.com/user-attachments/assets/3b55ef96-14ae-48c4-97e2-3a568722f99c)
![image](https://github.com/user-attachments/assets/f51fccc9-de14-4cef-92b7-2b277505b9de)

**Comment by akingscote** [comment]
Bumping to `3.90.0` **has resolved** the issue for me, thanks

**Comment by ziyeqf** [comment]
Hi all, 

Upgrading or adding `skip_provider_registration=true` both works for this issue. Just want to provide a complete solution for new comes:
1. upgrading resource provider to >= v3.90, as the fix on PR (https://github.com/hashicorp/terraform-provider-azurerm/pull/24571) was merged & release on that version.
2. If upgrading is unacceptable, adding `skip_provider_registration = true` as below also works
![image](https://github.com/user-attachments/assets/18f5695f-d4e2-431f-950a-1675a091dea0)
3. If do not want to add the feature as 2#, considering use the environment variable `ARM_SKIP_PROVIDER_REGISTRATION=true`, it acts as same as the feature as 2#

Then some comments on the original questions:
1. _I don't believe we use any resource of that provider! why do I see this error? How can I query if I have a resource under that provider?_
The Resource provider was defined in the required resource provider list (https://github.com/hashicorp/terraform-provider-azurerm/blob/157180ff56899e5844b47a6487ddd6c92ff963b9/internal/resourceproviders/required.go#L79), during the begining of Terraform execution, the provider will try to ensure all required providers be registed and will throw an error if it fails (before v3.90).
To find if any resourcees under the provider, all resources under this provider should starts with `azurerm_iot_time_series_insights_`, you can search it on your configurations.

2. _Can I fix all the jobs without needing to upgrade TF version in all repos (a lot)._
As others and the solution 2# and 3#, these work without uppgrading.

3. _Why can't I see any deprecation warning, so I could upgrade in a good time?_
Since you are not using these resources, the deprecation message was not shown... it might can be considered as a bug before v3.90, to fail the whole execution on resource provider registration failure.
For those who are using these resources, the deprecation message were added on v3.97 in this PR (https://github.com/hashicorp/terraform-provider-azurerm/pull/25312).

**Comment by archmangler** [comment]
Also seeing this issue intermittently at first and now consistently:

```
│ Error: Error ensuring Resource Providers are registered.
│ 
│ Terraform automatically attempts to register the Resource Providers it supports to
│ ensure it's able to provision resources.
│ 
│ If you don't have permission to register Resource Providers you may wish to use the
│ "skip_provider_registration" flag in the Provider block to disable this functionality.
│ 
│ Please note that if you opt out of Resource Provider Registration and Terraform tries
│ to provision a resource from a Resource Provider which is unregistered, then the errors
│ may appear misleading - for example:
│ 
│ > API version 2019-XX-XX was not found for Microsoft.Foo
│ 
│ Could indicate either that the Resource Provider "Microsoft.Foo" requires registration,
│ but this could also indicate that this Azure Region doesn't support this API version.
│ 
│ More information on the "skip_provider_registration" flag can be found here:
│ https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs#skip_provider_registration
│ 
│ Original Error: determining which Required Resource Providers require registration: the required Resource Provider "Microsoft.TimeSeriesInsights" wasn't returned from the Azure API
│ 
│   with provider["registry.terraform.io/hashicorp/azurerm"],
│   on local-providers.tf line 1, in provider "azurerm":
│    1: provider "azurerm" {
│ 
╵
```

This attempted workaround didn't work (in HK region):


```
+ az provider register --namespace Microsoft.TimeSeriesInsights
ERROR: (DisallowedProvider) The operation is not permitted for namespace 'Microsoft.TimeSeriesInsights'. List of permitted provider namespaces is .... <a long list of supported providers, none including TimeSeriesInsights>
```

**Comment by rcskosir** [comment]
Thank you for taking the time to raise this! I am going to close this with @ziyeqf‘s response as an answer. As noted above this has been resolved as of #24571. As such, I am going to mark this issue as closed.
